capa/features/extractors/binexport2/insn.py

```diff
@@ -0,0 +1,443 @@
+
+
+# end Ghidra symbol madness.
+#
+###############################################################################
+
+
+def extract_insn_number_features(
+    fh: FunctionHandle, _bbh: BBHandle, ih: InsnHandle
+) -> Iterator[Tuple[Feature, Address]]:
+    fhi: FunctionContext = fh.inner
+    ii: InstructionContext = ih.inner
+
+    be2 = fhi.ctx.be2
+    analysis = fhi.ctx.analysis
+
+    instruction_index = ii.instruction_index
+    instruction = be2.instruction[instruction_index]
+
+    _is_gsm = _is_ghidra_symbol_madness(be2, instruction_index)
+
+    for i, operand_index in enumerate(instruction.operand_index):
+        operand = be2.operand[operand_index]
+
+        if len(operand.expression_index) == 1 and _is_gsm:
+            # temporarily, we'll have to try to guess at the interpretation.
+            symbol = _gsm_get_instruction_operand(be2, instruction_index, i)
+
+            if symbol.startswith("#0x"):
+                # like:
+                # - type: SYMBOL
+                #   symbol: "#0xffffffff"
+                try:
+                    value = int(symbol[len("#") :], 0x10)
+                except ValueError:
+                    # failed to parse as integer
+                    continue
+
+                # handling continues below at label: has a value
+
+            elif symbol.startswith("0x"):
+                # like:
+                # - type: SYMBOL
+                #   symbol: "0x1000"
+                try:
+                    value = int(symbol, 0x10)
+                except ValueError:
+                    # failed to parse as integer
+                    continue
+
+                # handling continues below at label: has a value
+
+            else:
+                continue
+
+        elif len(operand.expression_index) == 2:
+            # from BinDetego,
+            # we get the following pattern for immediate constants:
+            #
+            # - type: SIZE_PREFIX
+            #   symbol: "b8"
+            # - type: IMMEDIATE_INT
+            #   immediate: 20588728364
+            #   parent_index: 0
+
+            expression0 = be2.expression[operand.expression_index[0]]
+            expression1 = be2.expression[operand.expression_index[1]]
+
+            if BinExport2.Expression.Type.SIZE_PREFIX != expression0.type:
+                continue
+
+            if BinExport2.Expression.Type.IMMEDIATE_INT != expression1.type:
+                continue
+
+            value = expression1.immediate
+
+            # handling continues below at label: has a value
+
+        else:
+            continue
+
+        # label: has a value
+
+        if analysis.base_address != 0x0:
+            # When the image is mapped at 0x0,
+            #  then its hard to tell if numbers are pointers or numbers.
+            # So be a little less conservative here.
+            if is_address_mapped(be2, value):
+                continue
+
+        if is_address_mapped(be2, value):
+            continue
+
```

Comment:
also exclude data references (for IDA we do this by checking for an offset https://github.com/mandiant/capa/blob/4b451566b2f782c18707f608d79324d36ba1e97a/capa/features/extractors/ida/insn.py#L139)